src/blueprints/tv2/tv2-on-timeline-generate-calculator.ts

```diff
@@ -0,0 +1,288 @@
+import { Tv2SisyfosPersistentLayerFinder } from './helpers/tv2-sisyfos-persistent-layer-finder'
+import { UnsupportedOperation } from '../../model/exceptions/unsupported-operation'
+import { Tv2BlueprintTimelineObject, Tv2PieceMetaData } from './value-objects/tv2-meta-data'
+import { Tv2MediaPlayer, Tv2StudioBlueprintConfiguration } from './value-objects/tv2-studio-blueprint-configuration'
+import { Timeline } from '../../model/entities/timeline'
+import { DeviceType } from '../../model/enums/device-type'
+import { AtemTimelineObjectContent, AtemType } from '../timeline-state-resolver-types/atem-types'
+import { A_B_SOURCE_LAYERS } from './value-objects/tv2-a-b-source-layers'
+import { Configuration } from '../../model/entities/configuration'
+import { Tv2BlueprintConfiguration } from './value-objects/tv2-blueprint-configuration'
+import { Tv2ShowStyleBlueprintConfiguration } from './value-objects/tv2-show-style-blueprint-configuration'
+import { SisyfosChannelsTimelineObject, SisyfosType } from '../timeline-state-resolver-types/sisyfos-types'
+import { OnTimelineGenerateResult } from '../../model/value-objects/on-timeline-generate-result'
+
+const ACTIVE_GROUP_PREFIX: string = 'active_group_'
+const LOOK_AHEAD_GROUP_ID: string = 'look_ahead_group'
+const PREVIOUS_GROUP_PREFIX: string = 'previous_group_'
+
+export class Tv2OnTimelineGenerateCalculator implements BlueprintOnTimelineGenerate {
+  constructor(private readonly sisyfosPersistentLayerFinder: Tv2SisyfosPersistentLayerFinder) {}
+
+  public onTimelineGenerate(
+    configuration: Configuration,
+    previousRundownPersistentState: RundownPersistentState,
+    activePart: Part,
+    previousPart: Part | undefined,
+    timeline: Timeline
+  ): OnTimelineGenerateResult {
+    const blueprintConfiguration: Tv2BlueprintConfiguration = {
+      studio: configuration.studio.blueprintConfiguration as Tv2StudioBlueprintConfiguration,
+      showStyle: configuration.showStyle.blueprintConfiguration as Tv2ShowStyleBlueprintConfiguration
+    }
+
+    const rundownPersistentState: Tv2RundownPersistentState = this.convertToTv2RundownPersistentState(previousRundownPersistentState)
+    const newRundownPersistentState: Tv2RundownPersistentState = {
+      activeMediaPlayerSessions: [],
+      isNewSegment: previousPart?.segmentId !== activePart.segmentId,
+    }
+
+    this.assignSisyfosPersistMetaData(newRundownPersistentState, activePart, previousPart, timeline)
+
+    newRundownPersistentState.activeMediaPlayerSessions = this.assignMediaPlayerSessions(rundownPersistentState.activeMediaPlayerSessions, timeline, blueprintConfiguration)
+
+    return { rundownPersistentState: newRundownPersistentState, timeline }
+  }
+
+  private convertToTv2RundownPersistentState(persistentState: RundownPersistentState): Tv2RundownPersistentState {
+    let rundownPersistentState: Tv2RundownPersistentState = persistentState as Tv2RundownPersistentState
+    if (!rundownPersistentState) {
```

Comment:
The very first time we calculate the `RundownPersistentState` we won't have a previous state since it is the first time, so yes, it can be potentially undefined...
(Also, when activating the Rundown we also start with a "fresh"/no previous state)

Reply:
Updated.